### DISCUSSION TITLE
Removing OBR and associated segments if no OBX

### DISCUSSION THREAD

**Discussion by rutledj** [nextgenhealthcare/connect#5244] [Q&A] [answered]
I'm going through and deleting some OBX segments based on a value in the OBX-5. Now I need to remove any OBR (and associated segments) if there are no OBXs left in that OBR block. Not sure of a good way to do that.

**Comment by jonbartels** [comment]
Show the code you're already using please.

Show sample inputs and outputs.

Please use code formatting so its easy to read.

Happy to help but give us something to chew on!

**Comment by rutledj** [comment]
Here is the code
```
        //Loop through the OBXs looking for 'Pending' in OBX-5 and remove them. 
	//If no OBXs are left then remove that OBR
 	for each(seg in msg..OBR)     
 	{		
		for each(obx in getSegmentsAfter(msg, msg.children()[seg.childIndex()], 'OBX',false, 'OBR'))
		{
			if(obx['OBX.5']['OBX.5.1'].toString().toUpperCase() == "PENDING")
			{
				delete msg.children()[obx.childIndex()];
			}
		}
}
```

Message may look like this:

```
MSH|^~\&|LAB|HMH||PVNMAHH|20220614163142-0400||ORU^R01|6535172.1|P|2.3|
PID|1|HMRL0262150||M000575624|DOE^JOHN||19381121|M||UN|
PV1|1|O|LAB.FAIR|EL|||CRUZCH^CRUZ^CHRISTINA|||||||FROM ACUTE||||REF||U|||||||||||||||||||HMH||REG|||202206140832|
OBR|1||4652508^LAB|UCREAT24^URINE CREATININE (24 HOUR)^L|||202206141030|||||||202206141133||1548813728^CRUZ^CHRISTINA^^^^^^NPI||03043726||||||LAB|F||^^^^^R||
NTE|1||BEGAN ON 6/13/22 AT 1030. ENDED 6/14/22 AT 1030.|
OBX|1|ST|UTV24^TOTAL VOLUME 24HRS,URINE^L|1|Pending|mL||||A^S|I|||||
OBX|2|ST|UCREA24^CREATININE 24 HOUR,URINE^L|1|Pending|g/24hr|0.98-2.2|||A^S|I|||||
NTE|1||PERFORMING SITE:Harrington Hospital|
NTE|1||100 South Street|
NTE|1||Southbridge, MA. 01550|
OBR|2||4652508^LAB|UCREAT24^URINE CREATININE (24 HOUR)^L|||202206141030|||||||202206141133||1548813728^CRUZ^CHRISTINA^^^^^^NPI||03043726||||||LAB|F||^^^^^R||
NTE|1||BEGAN ON 6/13/22 AT 1030. ENDED 6/14/22 AT 1030.|
OBX|1|ST|URCREATR^CREATININE,URINE RANDOM^L|1|39.70|mg/dL||||A^S|F|||202206141631||
```

I would want the first OBR and associated segments to be removed since both of the OBXs contain Pending.  The 2nd OBR and segments should stay.

**Reply by jonbartels** [reply]
What about doing this after you delete your OBXs?
```
if(getSegmentsAfter(msg, msg.children()[seg.childIndex()], 'OBX',false, 'OBR').length == 0){
   delete msg.children()[seg.childIndex()];
}
```

**Comment by rutledj** [comment]
Wouldn't that just delete the OBR segment and not the NTEs? Or would it get all of them?

**Reply by siddharthsharma9258** [reply]
OBR-NTE-OBX is a segment group, so deleting OBR should remove NTE as well. But that's considering you are using Strict parser.

**Comment by tonygermano** [comment, marked as answer]
This assumes it is possible for some OBX segments to be pending while others are not for the same OBR. If that can't happen, then both solutions can be simplified, as you don't need to delete obx segments separately from the entire order/observation group.

### Strict version
If your message conforms to the hl7 spec, as @siddharthsharma9258 mentioned, this can be done using the strict parser. It will require rewriting the code as the XML is completely different than the standard mirth parser due to it following the actual HL7 specification. You will no longer need `getSegmentsAfter`.

> WARNING! You may not want to use this as it does change the content of your OBX-10 field (which does not conform to the spec)

```js
for each (var orderObs in msg['ORU_R01.RESPONSE']['ORU_R01.ORDER_OBSERVATION']) {
	for each (var obs in orderObs['ORU_R01.OBSERVATION']) {
		if (obs['OBX']['OBX.5'].toString().toUpperCase() === 'PENDING') {
			// this will delete the OBSERVATION group,
			// including the OBX segment and any associated NTE segments
			delete obs[0]
		}
	}
	if (orderObs['ORU_R01.OBSERVATION']['OBX'].length() === 0) {
		// this will delete the ORDER_OBSERVATION group,
		// including the OBR segment and any associated NTE segments
		delete orderObs[0]
	}
}
```
### Standard version
This will do the same thing with `getSegmentsAfter` and the standard parser. Unlike the strict solution, it will not account for an optional ORC segment being present which does not appear in your sample message.

> Note that `getSegmentsAfter` returns a javascript array of xml objects, and this code treats the results as such. While `for each...in` loops will work on a javascript array most of the time, using them in this way is discouraged as it can produce unpredictable results in the same manner as using a `for...in` loop on an array. Not even existing in standard javascript, `for each...in` loops are intended to be used on xml lists and xml objects only.

```js
// delete javascript array of segments
function deleteSegList(list) {
	list.forEach(seg => {delete seg[0]})
}

// delete OBX segment and consecutive NTE segments
function deleteObservation(obx) {
	var nteList = getSegmentsAfter(msg, obx, 'NTE', true)
	deleteSegList(nteList)
	delete obx[0]
}

// delete OBR segment and all consecutive segments until the next OBR or end of message
function deleteOrderObservation(obr) {
	var allSegsUntilNextObr = getSegmentsAfter(msg, obr, /.*/)
	deleteSegList(allSegsUntilNextObr)
	delete obr[0]
}

for each (var obr in msg.OBR) {
	var obxList = getSegmentsAfter(msg, obr, 'OBX')
	var pending = obxList.filter(obx => obx['OBX.5']['OBX.5.1'].toString().toUpperCase() == "PENDING")
	if (obxList.length - pending.length === 0) {
		deleteOrderObservation(obr)
	} else if (pending.length > 0) {
		pending.forEach(deleteObservation)
	}
}
```

**Reply by siddharthsharma9258** [reply]
.filter() works now? So will .reduce()?

**Reply by pacmano1** [reply]
https://mozilla.github.io/rhino/compat/engines.html

**Reply by tonygermano** [reply]
Filter and reduce are ES5 functions, and ES5 has been well supported since before mirth 3.0.

**Reply by siddharthsharma9258** [reply]
Interesting. I never knew that.

**Reply by tonygermano** [reply]
@siddharthsharma9258  In older versions of mirth prior to the availability of ES6 mode, arrow functions were not supported, so it would need to be written as
```js
obxList.filter(function(obx) {return obx['OBX.5']['OBX.5.1'].toString().toUpperCase() == "PENDING"})
```

**Reply by Electro666** [reply]
If I use the 3 functions mentioned above with code listed below it will delete the OBR and OBXs.  How can I also delete the ORC above OBR?

for each (var obr in msg.OBR) {
	var obxList = getSegmentsAfter(msg, obr, 'OBX')
if (obr['OBR.4']['OBR.4.1'].toString() == '12345') {
	deleteOrderObservation(obr)
	}
}

**Reply by pacmano1** [reply]
@Electro666 create a new discussion thread, don't hijack old and answered ones please.  Just reference this thread on your new post.

Also post a complete before and after HL7 message - often helps others to see what you are talking about.

**Comment by rutledj** [comment]
Just curious about the getSegmentsAfter function.  For this code

```
// delete OBR segment and all consecutive segments until the next OBR or end of message
function deleteOrderObservation(obr) {
	var allSegsUntilNextObr = getSegmentsAfter(msg, obr, /.*/)
	deleteSegList(allSegsUntilNextObr)
	delete obr[0]
```
how does that stop at the next OBR when you don't specify a stop segment in the function call?

**Reply by tonygermano** [reply]
This is the official home of the `getSegmentsAfter` function: https://github.com/nextgenhealthcare/connect-examples/tree/master/Code%20Templates/Get%20Segments%20After%20a%20Particular%20Segment

This is the description of the 4th parameter:

> consecutiveInd - If true, indicates that the segments are expected to come directly after startSeg. If false, segments are collected until another segment with the same name as startSeg is encountered. Defaults to false.

I did not pass a 4th argument, so it defaulted to false, which will stop when it hits another segment with the same name as startSeg (or the end of the message.)

**Comment by siddharthsharma9258** [comment]
See : https://forums.mirthproject.io/forum/mirth-connect/support/7041-mirth-tools-user-defined-functions?p=52598#post52598
